c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -699,10 +854,27 @@ class DeviceCachingAllocator {
     // Free all non-split cached blocks
     free_blocks(large_blocks);
     free_blocks(small_blocks);
+
+    for (auto it = graph_pools.begin(); it != graph_pools.end(); ) {
+      // See notifyCaptureDestroy for the strategy here.
+      TORCH_INTERNAL_ASSERT(it->second.use_count >= 0);
+      if (it->second.use_count > 0) {
+        ++it;
+        continue;
+      }
+      free_blocks(it->second.small_blocks, &it->second);
+      free_blocks(it->second.large_blocks, &it->second);
+      if (it->second.cudaMalloc_count == 0) {
+        it = graph_pools.erase(it);
+      } else {
+        ++it;
```

Comment:
Err, when could this case possibly happen?

Reply:
The graph's private pool could still have blocks referenced by tensors the user holds, in which case free_blocks won't have cudaFreed all the pool's cudaMallocs yet. We can't erase the PrivatePool from graph_pools until the graph is destroyed _and_ the user drops all their references such that free_blocks finishes cleaning up the cudaMallocs. So free_blocks may only cudaFree most of the pool's memory back to the system, it might get called on the pool several times before all the memory is freed back to the system, depending how long the user holds onto references created during capture. In any case we shouldn't free anything the user's still referencing, those tensors might be results they're still using even after the graph is destroyed.

Reply:
At least, that was my plan. Do you see a problem here?

Reply:
I don't see a problem but I need to through and check the invariants again.